### ISSUE TITLE
Headless Snap RPC Does Not Work

### ISSUE THREAD

**Issue by joesixpack** [carsenk/denarius#267]
j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon getinfo
error: You must set rpcpassword=<password> in the configuration file:
/home/j/snap/denarius/common/.denarius/denarius.conf
If the file does not exist, create it with owner-readable-only file permissions.

j@cassandra:~/snap/denarius/common/.denarius$ nano denarius.conf

j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon --rpcpassword=test getinfo
error: incorrect rpcuser or rpcpassword (authorization failed)

**Comment by manosv** [comment]
you are doing it wrong.If you are actually in the same machine as the daemon you dont need to pass user or password when you invoke a command(assuming that you have rpcallowip=127.0.0.1 in denarius.conf)
denarius.daemon getinfo
if you are in a seperate machine then you need to pass rpcuser and rpcpassword
Important:When you make a change in denarius.conf you have to restart the wallet

**Comment by carsenk** [comment]
@joesixpack Denarius creates you a default denarius.conf, Running the daemon with `denarius.daemon` and then running `denarius.daemon stop` to shut down the wallet for your first run, then run `denarius.daemon` to start it and then `denarius.daemon getinfo` after starting it again or any RPC commands as they should work at that point as the denarius.conf would be filled in with random values and should work, The values must be saved into the .conf file and not passed via flags for the local RPC, it just needs RPC enabled to be able to run RPC commands.

**Comment by buzzkillb** [comment]
sample **denarius.conf**
```
rpcuser=denarius
rpcpassword=LONGPASSWORD
logtimestamps=1
rpcport=32369
port=33369
maxtargetupload=250
daemon=1
listen=1
server=1
```

**Comment by buzzkillb** [comment]
Something else that happens. If you start denarius.daemon with rpcuser and rpcpassword in denarius.conf,, let the daemon start up, and **then** change the rpcuser and rpcpassword in denarius.conf, you can lose control of the rpc and will have issues stopping the daemon with the _stop_ command. I think you need to change those when the wallet is not running.

**Comment by carsenk** [comment]
> Something else that happens. If you start denarius.daemon with rpcuser and rpcpassword in denarius.conf,, let the daemon start up, and **then** change the rpcuser and rpcpassword in denarius.conf, you can lose control of the rpc and will have issues stopping the daemon with the _stop_ command. I think you need to change those when the wallet is not running.

Yes that is correct, as the denarius.daemon command uses the RPC to connect to denariusd, you must shut down the node before changing any RPC settings in the .conf.

**Comment by joesixpack** [comment]
Still issues with this when managing denarius.daemon from snap-systemd.  Not sure if snap-systemd support is a new thing in 157 but 114 was still running and hogging the machine despite snap updates along the way up to 157.  My gut feel it is due to previously starting the daemon via the command line and not via snap-systemd.  But of course RPC only works that former way:

```
j@cassandra:~/snap/denarius/common/.denarius$ rm *.conf
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap start denarius.denariusd
Started.
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap stop denarius.denariusd
Stopped.
j@cassandra:~/snap/denarius/common/.denarius$ ls *.conf
ls: cannot access '*.conf': No such file or directory
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap start denarius.denariusd
Started.
j@cassandra:~/snap/denarius/common/.denarius$ ls *.conf
ls: cannot access '*.conf': No such file or directory
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap stop denarius.denariusd
Stopped.
j@cassandra:~/snap/denarius/common/.denarius$ ls *.conf
ls: cannot access '*.conf': No such file or directory
j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon
Denarius server starting
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap stop denarius.denariusd
Stopped.
j@cassandra:~/snap/denarius/common/.denarius$ ls *.conf
denarius.conf
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap start denarius.denariusd
Started.
j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon getinfo
error: couldn't connect to server
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap logs denarius.denariusd
2020-06-25T00:40:51Z systemd[1]: Starting Service for snap application denarius.denariusd...
2020-06-25T00:40:51Z systemd[1]: Started Service for snap application denarius.denariusd.
2020-06-25T00:42:10Z systemd[1]: Stopping Service for snap application denarius.denariusd...
2020-06-25T00:42:11Z denarius.denariusd[22127]: Denarius server starting
2020-06-25T00:42:11Z systemd[1]: Stopped Service for snap application denarius.denariusd.
2020-06-25T00:43:06Z systemd[1]: Starting Service for snap application denarius.denariusd...
2020-06-25T00:43:06Z systemd[1]: Started Service for snap application denarius.denariusd.
2020-06-25T00:43:06Z denarius.denariusd[22395]: Denarius: Unable to bind to 0.0.0.0:33369 on this computer. Denarius is probably already running.
2020-06-25T00:43:06Z denarius.denariusd[22395]: Denarius: Failed to listen on any port. Use -listen=0 if you want this.
2020-06-25T00:43:06Z denarius.denariusd[22395]: Denarius server starting
j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon getinfo
error: couldn't connect to server
j@cassandra:~/snap/denarius/common/.denarius$ sudo systemctl status snap.denarius.denariusd.service
● snap.denarius.denariusd.service - Service for snap application denarius.denariusd
   Loaded: loaded (/etc/systemd/system/snap.denarius.denariusd.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Thu 2020-06-25 02:43:06 CEST; 47s ago
  Process: 22395 ExecStart=/usr/bin/snap run denarius.denariusd (code=exited, status=0/SUCCESS)
 Main PID: 22435 (code=exited, status=0/SUCCESS)

Jun 25 02:43:06 cassandra systemd[1]: Starting Service for snap application denarius.denariusd...
Jun 25 02:43:06 cassandra systemd[1]: Started Service for snap application denarius.denariusd.
Jun 25 02:43:06 cassandra denarius.denariusd[22395]: Denarius: Unable to bind to 0.0.0.0:33369 on this computer. Denarius is probably already runn
Jun 25 02:43:06 cassandra denarius.denariusd[22395]: Denarius: Failed to listen on any port. Use -listen=0 if you want this.
Jun 25 02:43:06 cassandra denarius.denariusd[22395]: Denarius server starting
j@cassandra:~/snap/denarius/common/.denarius$ ps -e | grep denarius
22328 ?        00:02:05 denariusd
j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon stop
Denarius server stopping, please wait a few minutes for full shutdown...
j@cassandra:~/snap/denarius/common/.denarius$ denariusd: /build/denarius/parts/daemon/install/usr/include/boost/thread/pthread/recursive_mutex.hpp:113: void boost::recursive_mutex::lock(): Assertion `!pthread_mutex_lock(&m)' failed.

j@cassandra:~/snap/denarius/common/.denarius$ ps -e | grep denarius
j@cassandra:~/snap/denarius/common/.denarius$ ls *.conf
denarius.conf
j@cassandra:~/snap/denarius/common/.denarius$ sudo snap start denarius.denariusd
Started.
j@cassandra:~/snap/denarius/common/.denarius$ ps -e | grep denarius
22783 ?        00:00:01 denariusd
j@cassandra:~/snap/denarius/common/.denarius$ denarius.daemon getinfo
error: incorrect rpcuser or rpcpassword (authorization failed)
j@cassandra:~/snap/denarius/common/.denarius$
```

**Comment by carsenk** [comment]
So there is a Snap daemon and a daemon of Denarius, you should be trying to run the daemon of Denarius with denarius.daemon to start it and denarius.daemon to stop it, denarius.daemon <your command> not denarius.denariusd

You can ensure Snap stops all previous running Snap Daemons (Related to Snap not Denarius)

```
snap stop <snap>[.<app>] – Stops all daemons in the snap, or only the selected one if specified.
snap start <snap>[.<app>] – Starts all daemons in the snap, or only the selected one if specified.
snap restart <snap>[.<app>] – Restarts all daemons in the snap, or only the selected one if specified. Daemons which are not yet running will be started as well.
snap restart --reload <snap>[.<app>] – Reloads all daemons in the snap, or only the selected one if specified. Daemons that do not support reloading are restarted instead as long as they were already running (systemd’s try-reload-or-restart).
snap log <snap>[.<app>] – Shows logs for all daemons in the snap, or only the selected one if specified.
snap status <snap>[.<app>] – Shows human-oriented status summary for all daemons in the snap, or only the selected one if specified (systemd’s status).
```

**Comment by buzzkillb** [comment]
Another thing to try is stop the daemon. Then remove denarius.conf from ~/snap/denarius/common/.denarius/denarius.conf

Then restart the daemon `denarius.daemon` , this will regenerate a clean denarius.conf that should work.

I have used the snap QT/daemon combo on Arch, Manjaro, Ubuntu, and Debian so far and its been fine. I tried Fedora on Qubes OS, and had an issue where I could not update the snap, but the daemon and QT still worked fine.
